### ISSUE TITLE
Better support for custom color and fill aesthetics?

### ISSUE THREAD

**Issue by clauswilke** [tidyverse/ggplot2#2345]
Hello,

I'm wondering whether it would be possible to better support custom color and fill aesthetics. What I mean are cases where maybe we want two separate color mappings in the same geom, e.g. `line_color` and `point_color`. With the current ggplot2, if I create such aesthetics, I can't take advantage of any of the existing color scales. But with a few small changes, I think this would be possible.

First, to demonstrate that this is a real situation, some example code from ggridges:
```r
cols = c("#E69F00", "#56B4E9", "#009E73")
cols_dark = c('#C18511', '#289ACF', '#0D8561')
ggplot(iris, aes(x=Sepal.Length, y=Species, fill = Species)) +
  geom_density_ridges(aes(point_color = Species), alpha = .2, jittered_points = TRUE) +
  scale_fill_manual(values = cols, guide = "none") + scale_y_discrete(expand = c(0.01, 0)) +
  ggplot2:::manual_scale("point_color", values = cols_dark, guide = "none") +
  theme_ridges(grid = FALSE, center = TRUE)
```
<img width="699" alt="screen shot 2017-11-18 at 3 20 39 pm" src="https://user-images.githubusercontent.com/4210929/32985579-1f43796c-cc84-11e7-8432-f40c59bbf9a6.png">

In this example, everything needs to be drawn at once in one geom, because the filled areas and solid lines can overlap the points. But, I want to be able to specify line colors and point colors separately, and the point colors should also be different from the fill colors. The only way I see to solve the problem is to add a `point_color` aesthetic.

My code example brings up one immediate and easy to fix issue: ggplot2 doesn't export `manual_scale`, but it's what is needed here to make this example work without too much effort. Second, it would be helpful if the various existing color scales took an optional `aesthetics` argument that can be used to override the aesthetics in the function name. While this may sound a bit strange, the resulting code would be quite legible and look reasonable. For example, using the default hue color scale for fill and also applying it to the points might look like this:

```r
ggplot(iris, aes(x=Sepal.Length, y=Species, fill = Species)) +
  geom_density_ridges(aes(point_color = Species), alpha = .2, jittered_points = TRUE) +
  scale_color_hue(aesthetics = "point_color") + scale_y_discrete(expand = c(0.01, 0)) +
  theme_ridges(grid = FALSE, center = TRUE)
```

So, in summary, I'm making the following requests:
1. Export `manual_scale`
2. Add optional `aesthetics` parameter to all existing color scales

I'd be happy to work on a pull request if you're willing to consider these changes.

**Comment by clauswilke** [comment]
Some more thoughts on this: I now think all scales should have an optional aesthetics parameter, not just color scales. The same issue exists, e.g., with `scale_shape()`. And I don't think it would make sense to implement separate scales functions that don't contain the aesthetic in their name, as in `scale_hue()`, because then one quickly runs into trouble. E.g., the generic version of `scale_shape()` would be `scale()`, which obviously doesn't make sense.

I think it's important to realize that the word "color" or "shape" or "size" really serves two distinct purposes in the scale names: First, it defines what type of a scale we're dealing with (a color scale, or a shape scale, etc). Second, it names the specific aesthetic for which this scale works. Those two purposes are currently tightly coupled in ggplot2, but they don't have to be.

For example, it would be perfectly reasonable to always implement color scales first as `scale_color_*` and  then provide fill scales simply by calling the color version with `aesthetics = "fill"`, as in:
```r
scale_fill_hue <- function(...) scale_color_hue(..., aesthetics = "fill")
```

**Comment by clauswilke** [comment]
After having worked with this for about a day now, I have found one other issue: The default guide function for continuous color scales, `guide_colorbar`, does not want to work with aesthetics it doesn't know. The `guide_legend` function does not have this limitation. Thus, I have been able to draw legends for aesthetics `point_size`, `point_shape`, and discrete `point_color` and `point_fill`, but not for continuous `point_color` or `point_fill`.

**Comment by clauswilke** [comment]
I'm closing this because main primary issues have been addressed with https://github.com/tidyverse/ggplot2/pull/2555. I'll open a new issue to collect thoughts for future directions.
